### ISSUE TITLE
Add Python client for direct training data access

### ISSUE THREAD

**Issue by jbarnes850** [Arc-Computer/atlas-sdk#120]
## Summary

The current JSONL export workflow forces a two-hop serialization path (Postgres → JSONL → Python objects) that introduces schema drift, performance overhead, and silent data loss when training models in ATLAS Core. This issue proposes exposing SDK's database models directly via a Python client, enabling ATLAS Core to query training data without intermediate file exports.

## Problem Statement

**Current Architecture:**
```
Runtime → Postgres → JSONL Export → File System → ATLAS Core Loader → Training
```

**Critical Issues:**

1. **Schema Drift**: SDK and ATLAS Core maintain parallel schemas that inevitably diverge (see [ATLAS#38](https://github.com/Arc-Computer/ATLAS/issues/38)). When SDK adds fields like `artifacts`, `deliverable`, `session_reward`, etc., ATLAS Core silently drops them until manually updated.

2. **Performance Overhead**: Large JSONL files require loading entire datasets into memory. No incremental loading, no filtering at source, no columnar access.

3. **Silent Data Loss**: New telemetry fields added to SDK exports are dropped by ATLAS Core loader, breaking learning measurement pipelines that expect `learning_history`, `trajectory_events`, `adaptive_summary`, etc.

4. **No Query Capabilities**: Training workflows need filtered queries (e.g., `min_reward >= 0.8`, `date > last_checkpoint`, `learning_key = 'specific_task'`). Current approach requires loading everything to filter anything.

5. **Maintenance Burden**: Every SDK schema change requires coordinated updates across:
   - SDK export logic (`atlas/cli/jsonl_writer.py`)
   - SDK schema dataclasses (`atlas/runtime/schema.py`)
   - ATLAS Core schema dataclasses (`atlas_core/runtime/schema.py`)
   - ATLAS Core coercion functions (`trainers/runtime_dataset.py`)
   - Test fixtures in both repos

## Proposed Solution

**Add `atlas.training_data` module for direct database access:**

```python
# atlas/training_data/__init__.py
from atlas.database import get_session, SessionModel, StepModel
from atlas.runtime.schema import AtlasSessionTrace
from sqlalchemy import select
from typing import List, Optional

def get_training_sessions(
    db_url: str,
    min_reward: float = 0.0,
    limit: int = 1000,
    offset: int = 0,
    created_after: Optional[datetime] = None,
    learning_key: Optional[str] = None,
) -> List[AtlasSessionTrace]:
    """
    Query training sessions directly from Postgres.
    
    Args:
        db_url: Postgres connection string
        min_reward: Minimum final reward threshold
        limit: Maximum sessions to return
        offset: Pagination offset
        created_after: Filter sessions after this timestamp
        learning_key: Filter by learning key
        
    Returns:
        List of AtlasSessionTrace objects with all telemetry preserved
    """
    session = get_session(db_url)
    
    query = select(SessionModel).where(SessionModel.final_reward >= min_reward)
    
    if created_after:
        query = query.where(SessionModel.created_at >= created_after)
    
    if learning_key:
        query = query.where(
            SessionModel.session_metadata['learning_key'].astext == learning_key
        )
    
    query = query.order_by(SessionModel.created_at.desc()).limit(limit).offset(offset)
    
    sessions = session.execute(query).scalars().all()
    return [session_model.to_atlas_trace() for session_model in sessions]


def get_session_by_id(db_url: str, session_id: str) -> Optional[AtlasSessionTrace]:
    """Retrieve specific session by ID."""
    session = get_session(db_url)
    session_model = session.query(SessionModel).filter_by(id=session_id).first()
    return session_model.to_atlas_trace() if session_model else None


def count_training_sessions(
    db_url: str,
    min_reward: float = 0.0,
    created_after: Optional[datetime] = None,
) -> int:
    """Count sessions matching criteria without loading data."""
    session = get_session(db_url)
    query = select(func.count(SessionModel.id)).where(
        SessionModel.final_reward >= min_reward
    )
    if created_after:
        query = query.where(SessionModel.created_at >= created_after)
    return session.execute(query).scalar()
```

**New Architecture:**
```
Runtime → Postgres ← ATLAS Core (direct query) → Training
                  ↓
            JSONL Export (optional, archival only)
```

### 1. Eliminates Schema Drift
- **Single source of truth**: Postgres schema IS the schema
- SDK's `SessionModel.to_atlas_trace()` method owns serialization
- ATLAS Core uses SDK's conversion logic directly
- Impossible for schemas to diverge

### 2. Performance Improvements
- **Incremental loading**: Query only new sessions since last checkpoint
- **Filtered queries**: `WHERE reward > 0.8 AND created_at > '2024-01-01'`
- **Columnar access**: Load only needed fields (future optimization)
- **Pagination**: Process large datasets in chunks without memory overflow

### 3. Enterprise Privacy & Security
- **On-premises deployment**: All data remains on customer infrastructure
- **Database access controls**: Standard Postgres auth, roles, audit logs
- **Encryption**: Native Postgres encryption at rest and in transit
- **Compliance**: Postgres has SOC2, HIPAA, ISO 27001 certifications
- **Air-gap capable**: No external dependencies, works fully offline
- **Better than flat files**: Centralized access control vs filesystem permissions

### 4. Learning Measurement Pipeline (Phase 1)
Enables direct queries for five critical metrics:

- **Learning Cue Hit Rate**: Query `learning_history`, `trajectory_events`, `guidance`
- **Adoption Rate**: Query `student_learning`, `teacher_learning`, `attempt_history`
- **Reward Delta**: Query `session_reward`, reward breakdowns, learning history
- **Token Delta**: Query `runtime`, `artifacts`, `deliverable`, attempt history
- **Transfer Success**: Query `adaptive_summary`, `learning_key`, `triage_dossier`

All fields accessible without maintaining parallel schemas or re-parsing JSONL.

### 5. Natural Cloud Migration Path

**Phase 1 (Open Source - Now):**
```python
# On-prem: Direct local Postgres connection
sessions = get_training_sessions("postgresql://localhost/atlas")
```

**Phase 2 (Hybrid - Future):**
```python
# Add REST/gRPC wrapper over Postgres queries
sessions = get_training_sessions("http://api.atlas.internal/training")
```

**Phase 3 (Managed Service - Optional):**
```python
# Arc-hosted API for customers who want managed service
sessions = get_training_sessions("https://api.arc.dev/training", api_key=key)
```

Same client interface, different backend. On-prem customers stay on-prem, cloud customers migrate when ready.

### Phase 1: Core Client (This Issue)

1. **Create `atlas/training_data/` module**
   - `client.py`: Core query functions
   - `filters.py`: Query filter helpers
   - `pagination.py`: Cursor-based pagination
   - `__init__.py`: Public API exports

2. **Add `SessionModel.to_atlas_trace()` method**
   - Convert SQLAlchemy model to `AtlasSessionTrace` dataclass
   - Include ALL fields (no silent data loss)
   - Handle JSON field deserialization (`session_metadata`, etc.)

3. **Add `StepModel.to_atlas_step_trace()` method**
   - Convert SQLAlchemy model to `AtlasStepTrace` dataclass
   - Preserve `artifacts`, `deliverable`, metadata
   - Handle nested structures (reward breakdown, validation, etc.)

4. **Update ATLAS Core** ([ATLAS#38](https://github.com/Arc-Computer/ATLAS/issues/38))
   - Add `atlas_core/loaders/postgres_loader.py`
   - Use `atlas.training_data` client for primary path
   - Keep JSONL loader for backward compatibility

5. **Documentation**
   - Enterprise deployment guide emphasizing privacy/security
   - Query examples for common training workflows
   - Migration guide from JSONL to direct access

6. **Testing**
   - Unit tests: Query filtering, pagination, type conversion
   - Integration tests: ATLAS Core can load sessions without data loss
   - Performance benchmarks: Compare to JSONL loading

### Phase 2: Advanced Features (Future)

- Query optimizer for common patterns
- Streaming results for large datasets
- Connection pooling for multi-process training
- Read replicas support for high-throughput training

### Phase 3: API Layer (Future)

- REST/gRPC wrapper over training data queries
- Authentication and rate limiting
- Multi-tenancy support for managed service

## Backward Compatibility

**JSONL export remains supported:**
- Keep `arc-atlas export` command for archival/sharing
- Useful for data transfer between environments
- Human-readable format for debugging
- NOT the primary training path

**ATLAS Core supports both:**
- New path: Direct Postgres queries (recommended)
- Legacy path: JSONL loading (deprecated, maintained for compatibility)

## Success Criteria

- [ ] ATLAS Core can query training sessions without JSONL export
- [ ] All SDK fields accessible without schema drift risk
- [ ] Query performance: Filter 1M sessions to 10K in <1 second
- [ ] Enterprise deployment: Works air-gapped, on-prem
- [ ] Documentation: Clear migration path from JSONL
- [ ] Tests: 100% field preservation verified
- [ ] Backward compatibility: JSONL export still works

## Related Issues

- [ATLAS#38](https://github.com/Arc-Computer/ATLAS/issues/38): Align runtime trace schema (stopgap fix for JSONL approach)
- This issue provides the long-term architectural solution

## Questions for Discussion

1. Should we expose raw SQLAlchemy models or only through `training_data` client?
2. Should we add query builder DSL or stick with function parameters?
3. How should we handle connection pooling for distributed training?
4. Should Phase 1 include basic caching layer?

---

**Context**: This aligns with Phase 1 product vision (execute runtime to maximize inference-time learning) by unblocking learning measurement pipeline and Phase 2 data curation experiments (GRPO training, distillation).
